File: src/templates/skills/api-pagination/templates/rails/cursor_pagination.rb
# frozen_string_literal: true

module Pagination
  # Cursor-based pagination for efficient navigation through large datasets.
  # Uses opaque cursors to track position without offset calculations.
  #
  # Best for: Large datasets, infinite scroll, real-time data, mobile apps
  # Advantages: Consistent performance, handles data changes gracefully
  #
  # Example usage in controller:
  #
  #   class UsersController < ApplicationController
  #     include PaginationConcern
  #
  #     def index
  #       users = User.active.order(:id)
  #       result = cursor_paginate(users, params[:cursor], params[:page_size] || {{defaultPageSize}})
  #
  #       render json: result
  #     end
  #   end
  module CursorPagination
    DEFAULT_PAGE_SIZE = {{defaultPageSize}}
    MAX_PAGE_SIZE = {{maxPageSize}}

    # Encode a value as a base64 cursor
    #
    # @param value [Object] The value to encode
    # @return [String] Base64 encoded cursor
    def encode_cursor(value)
      return '' if value.nil?

      Base64.strict_encode64(value.to_s)
    end

    # Decode a base64 cursor
    #
    # @param cursor [String] The cursor to decode
    # @return [String] Decoded cursor value
    def decode_cursor(cursor)
      return '' if cursor.blank?

      Base64.strict_decode64(cursor)
    rescue ArgumentError
      ''
    end

    # Paginate an ActiveRecord relation using cursor-based pagination
    #
    # @param relation [ActiveRecord::Relation] The query to paginate
    # @param cursor [String, nil] Current cursor (nil for first page)
    # @param page_size [Integer] Number of items per page
    # @param cursor_field [Symbol] Field to use as cursor (default: :id)
    # @param direction [Symbol] Sort direction (:asc or :desc)
    # @return [Hash] Paginated result with items and metadata
    #
    # @example
    #   result = cursor_paginate(User.active, params[:cursor], 20, :id, :asc)
    #   # => {
    #   #   items: [...],
    #   #   pagination: {
    #   #     next_cursor: "...",
    #   #     previous_cursor: "...",
    #   #     has_next: true,
    #   #     has_previous: false,
    #   #     page_size: 20
    #   #   }
    #   # }
    def cursor_paginate(relation, cursor = nil, page_size = DEFAULT_PAGE_SIZE, cursor_field = :id, direction = :asc)
      # Constrain page size
      page_size = [[page_size.to_i, 1].max, MAX_PAGE_SIZE].min

      # Apply cursor filter if provided
      if cursor.present?
        decoded_cursor = decode_cursor(cursor)

        relation = if direction == :asc
                     relation.where("#{cursor_field} > ?", decoded_cursor)
                   else
                     relation.where("#{cursor_field} < ?", decoded_cursor)
                   end
      end

      # Ensure ordering
      relation = relation.order(cursor_field => direction)

      # Fetch one extra item to check for next page
      items = relation.limit(page_size + 1).to_a

      has_next = items.size > page_size
      items = items.take(page_size) if has_next

      # Generate cursors
      next_cursor = if has_next && items.any?
                      encode_cursor(items.last.send(cursor_field))
                    end

      previous_cursor = if cursor.present? && items.any?
                          encode_cursor(items.first.send(cursor_field))
                        end

      {
        items: items,
        pagination: {
          next_cursor: next_cursor,
          previous_cursor: previous_cursor,
          has_next: has_next,
          has_previous: cursor.present?,
          page_size: page_size
        }
      }
    end

    # Cursor paginate and render as JSON with custom serializer
    #
    # @param relation [ActiveRecord::Relation] The query to paginate
    # @param cursor [String, nil] Current cursor
    # @param page_size [Integer] Number of items per page
    # @param serializer [Class] Serializer class to use
    # @param cursor_field [Symbol] Field to use as cursor
    # @param direction [Symbol] Sort direction
    # @return [Hash] Serialized paginated response
    def cursor_paginate_json(relation, cursor, page_size, serializer: nil, cursor_field: :id, direction: :asc)
      result = cursor_paginate(relation, cursor, page_size, cursor_field, direction)

      items = if serializer
                result[:items].map { |item| serializer.new(item).as_json }
              else
                result[:items].as_json
              end

      {
        data: items,
        pagination: result[:pagination]
      }
    end
  end
end

File: src/templates/skills/api-pagination/templates/aspnet-core/CursorPagination.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Microsoft.EntityFrameworkCore;

namespace {{namespace}}.Pagination
{
    /// <summary>
    /// Cursor-based pagination for efficient navigation through large datasets.
    /// Uses opaque cursors to track position without offset calculations.
    /// </summary>
    /// <remarks>
    /// Best for: Large datasets, infinite scroll, real-time data, mobile apps
    /// Advantages: Consistent performance, handles data changes gracefully
    /// </remarks>
    public class CursorPagination<T> where T : class
    {
        /// <summary>
        /// The items for the current page
        /// </summary>
        public IReadOnlyList<T> Items { get; }

        /// <summary>
        /// Cursor for the next page (null if no more items)
        /// </summary>
        public string? NextCursor { get; }

        /// <summary>
        /// Cursor for the previous page (null if on first page)
        /// </summary>
        public string? PreviousCursor { get; }

        /// <summary>
        /// Whether there are more items after this page
        /// </summary>
        public bool HasNext { get; }

        /// <summary>
        /// Whether there are items before this page
        /// </summary>
        public bool HasPrevious { get; }

        /// <summary>
        /// Number of items per page
        /// </summary>
        public int PageSize { get; }

        public CursorPagination(
            IReadOnlyList<T> items,
            string? nextCursor,
            string? previousCursor,
            bool hasNext,
            bool hasPrevious,
            int pageSize)
        {
            Items = items;
            NextCursor = nextCursor;
            PreviousCursor = previousCursor;
            HasNext = hasNext;
            HasPrevious = hasPrevious;
            PageSize = pageSize;
        }
    }

    /// <summary>
    /// Helper methods for cursor-based pagination
    /// </summary>
    public static class CursorPaginationHelper
    {
        /// <summary>
        /// Encode a value as a base64 cursor
        /// </summary>
        public static string EncodeCursor(object value)
        {
            if (value == null) return string.Empty;

            var bytes = Encoding.UTF8.GetBytes(value.ToString() ?? string.Empty);
            return Convert.ToBase64String(bytes);
        }

        /// <summary>
        /// Decode a base64 cursor to a string value
        /// </summary>
        public static string DecodeCursor(string cursor)
        {
            if (string.IsNullOrEmpty(cursor)) return string.Empty;

            try
            {
                var bytes = Convert.FromBase64String(cursor);
                return Encoding.UTF8.GetString(bytes);
            }
            catch
            {
                return string.Empty;
            }
        }

        /// <summary>
        /// Paginate an IQueryable using cursor-based pagination
        /// </summary>
        /// <typeparam name="T">Entity type</typeparam>
        /// <typeparam name="TKey">Key type for cursor</typeparam>
        /// <param name="query">The base query</param>
        /// <param name="keySelector">Function to select the cursor key</param>
        /// <param name="cursor">Current cursor (null for first page)</param>
        /// <param name="pageSize">Number of items per page</param>
        /// <param name="ascending">Sort direction (true for ascending, false for descending)</param>
        /// <returns>Paginated result</returns>
        /// <example>
        /// <code>
        /// // In your controller:
        /// [HttpGet]
        /// public async Task&lt;ActionResult&lt;CursorPagination&lt;User&gt;&gt;&gt; GetUsers(
        ///     [FromQuery] string? cursor = null,
        ///     [FromQuery] int pageSize = {{defaultPageSize}})
        /// {
        ///     var result = await CursorPaginationHelper.PaginateAsync(
        ///         _context.Users.AsQueryable(),
        ///         u => u.Id,
        ///         cursor,
        ///         pageSize
        ///     );
        ///
        ///     return Ok(result);
        /// }
        /// </code>
        /// </example>
        public static async Task<CursorPagination<T>> PaginateAsync<T, TKey>(
            IQueryable<T> query,
            Func<T, TKey> keySelector,
            string? cursor,
            int pageSize,
            bool ascending = true) where T : class where TKey : IComparable<TKey>
        {
            pageSize = Math.Min(pageSize, {{maxPageSize}});

            // Apply cursor filter if provided
            if (!string.IsNullOrEmpty(cursor))
            {
                var decodedCursor = DecodeCursor(cursor);

                // This is a simplified version - in production, you'd want to handle type conversion
                if (typeof(TKey) == typeof(int) && int.TryParse(decodedCursor, out var intCursor))
                {
                    query = ascending
                        ? query.Where(x => ((IComparable)keySelector(x)).CompareTo(intCursor) > 0)
                        : query.Where(x => ((IComparable)keySelector(x)).CompareTo(intCursor) < 0);
                }
                else if (typeof(TKey) == typeof(Guid) && Guid.TryParse(decodedCursor, out var guidCursor))
                {
                    query = ascending
                        ? query.Where(x => ((IComparable)keySelector(x)).CompareTo(guidCursor) > 0)
                        : query.Where(x => ((IComparable)keySelector(x)).CompareTo(guidCursor) < 0);
                }
            }

            // Fetch one extra item to check if there's a next page
            var items = await query
                .Take(pageSize + 1)
                .ToListAsync();

            var hasNext = items.Count > pageSize;
            if (hasNext)
            {
                items = items.Take(pageSize).ToList();
            }

            var nextCursor = hasNext && items.Any()
                ? EncodeCursor(keySelector(items.Last()))
                : null;

            var previousCursor = !string.IsNullOrEmpty(cursor) && items.Any()
                ? EncodeCursor(keySelector(items.First()))
                : null;

            return new CursorPagination<T>(
                items.AsReadOnly(),
                nextCursor,
                previousCursor,
                hasNext,
                !string.IsNullOrEmpty(cursor),
                pageSize
            );
        }
    }
}
